File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedPositionXYZ.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.Player;
import cn.nukkit.level.Location;
import cn.nukkit.math.Mth;
import cn.nukkit.math.Vector3;
import cn.nukkit.utils.BinaryStream;
import cn.nukkit.utils.TextFormat;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.playback.PlayerPlaybackEngine;
import net.easecation.ghosty.recording.player.PlayerRecordNode;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 15:22.
 */
public class PlayerUpdatedPositionXYZ implements PlayerUpdated {
    private double x;
    private double y;
    private double z;

    public static PlayerUpdatedPositionXYZ of(double x, double y, double z) {
        return new PlayerUpdatedPositionXYZ(x, y, z);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_POSITION_XYZ;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        Location location = ghost.getLocation();
        location.x = x;
        location.y = y;
        location.z = z;
        ghost.teleport(location);
        PlayerPlaybackEngine engine = ghost.getEngine();
        if (engine == null) {
            return;
        }
        if (ghost.lastPosition == null) {
            ghost.lastPosition = location;
            ghost.lastMoveTick = engine.getTick();
        } else {
            if (engine.displayMovingSpeed && engine.getTick() % 10 == 0) {
                // 计算距离
                double distance = location.distance(ghost.lastPosition) / (engine.getTick() - ghost.lastMoveTick) * 20;
                if (distance >= 0.01) {
                    String distanceStr = String.valueOf(Mth.round(distance, 3));
                    // 显示速度（包含双方的Ping数据）
                    String msg = "[Moving] " + distanceStr + TextFormat.WHITE + " "
                        + "[" + PlayerUpdatedPing.getDisplayPing(ghost.lastPing) + TextFormat.WHITE + "]" + ghost.getAliasName();
                    for (Player viewer : ghost.getViewers().values()) {
                        viewer.sendMessage(msg);
                    }
                }
            }

            ghost.lastPosition = location;
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setX(x);
        node.setY(y);
        node.setZ(z);
        return node;
    }

    public PlayerUpdatedPositionXYZ(BinaryStream stream) {
        this.read(stream);
    }

    private PlayerUpdatedPositionXYZ(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getZ() {
        return z;
    }

    public Vector3 asVector3() {
        return new Vector3(this.x, this.y, this.z);
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedPositionXYZ o)) return false;
        return (x == o.x) && (y == o.y) && (z==o.z);
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putFloat((float) this.x);
        stream.putFloat((float) this.y);
        stream.putFloat((float) this.z);
    }

    @Override
    public void read(BinaryStream stream) {
        this.x = stream.getFloat();
        this.y = stream.getFloat();
        this.z = stream.getFloat();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedPositionXYZ{" +
            "x=" + x +
            ", y=" + y +
            ", z=" + z +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedPositionXYZ.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.math.Mth
import cn.nukkit.math.Vector3
import cn.nukkit.utils.BinaryStream
import cn.nukkit.utils.TextFormat
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.player.PlayerRecordNode

@Serializable
data class PlayerUpdatedPositionXYZ(
    val x: Double,
    val y: Double,
    val z: Double,
) : PlayerUpdated, UpdateWithState {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_POSITION_XYZ

    override fun processTo(ghost: PlaybackNPC) {
        ghost.teleport(ghost.location.setComponents(x, y, z))
        val engine = ghost.engine ?: return
        if (ghost.lastPosition != null) {
            if (engine.displayMovingSpeed && engine.tick % 10 == 0) {
                val delta = engine.tick - ghost.lastMoveTick
                val distance = ghost.location.distance(ghost.lastPosition) / delta * 20
                if (distance >= 0.01) {
                    val msg = buildString {
                        append("[Moving] ${Mth.round(distance, 3)}${TextFormat.WHITE} ")
                        append("[${PlayerUpdatedPing.getDisplayPing(ghost.lastPing)}${TextFormat.WHITE}]${ghost.aliasName}")
                    }
                    ghost.viewers.values.forEach { it.sendMessage(msg) }
                }
            }
            ghost.lastPosition = ghost.location
        } else {
            ghost.lastPosition = ghost.location
            ghost.lastMoveTick = engine.tick
        }
    }

    override fun applyTo(node: PlayerRecordNode) = node.also {
        it.x = this.x
        it.y = this.y
        it.z = this.z
    }

    fun asVector3() = Vector3(x, y, z)

    companion object {
        @JvmStatic
        fun of(x: Double, y: Double, z: Double) = PlayerUpdatedPositionXYZ(x, y, z)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedPositionXYZ> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedPositionXYZ> {
        override fun write(updated: PlayerUpdatedPositionXYZ, stream: BinaryStream) {
            stream.putFloat(updated.x.toFloat())
            stream.putFloat(updated.y.toFloat())
            stream.putFloat(updated.z.toFloat())
        }

        override fun read(stream: BinaryStream): PlayerUpdatedPositionXYZ {
            return PlayerUpdatedPositionXYZ(
                x = stream.float.toDouble(),
                y = stream.float.toDouble(),
                z = stream.float.toDouble(),
            )
        }
    }
}
